Add explicit no-pie arguments

dcbaker opened the pull request:
We're relying on GCC's no-pie by default implementation, which isn't a good plan. especially given that macOS is pie by default. This adds the necessary explicit flags for GCC, clang, and ICC to turn pie on and off. I've only tested this on macOS and Linux, so it may well be broken on windows.

Additionally, this allows pie to be set (but does nothing) for languages that don't support it. Which is required to be able to mix languages that have PIE and those that don't. I'm not sure if that's a good idea or not.

bwidawsk commented:
>  relying on GCC's no-pie by default implementation, which isn't a good plan. especially given that macOS is pie by default. This adds the necessary explicit flags for GCC, clang, and ICC to turn pie on and off. I've only test

Fixes #4651

dcbaker commented:
@xclaesse, I think you wrote the original pie stuff. Does this look good to you?

xclaesse commented:
Hmm, I'm wondering if it should still be using platform/compiler's default if pie is not explicitly set to true/false. This change means we won't be PIE any more on MacOSX by default if I understand correctly, which is not backward compatible change, right?

jpakkane commented:
If that is the case then either the default value should be `true`, or the option should be a tristate of `true`, `false`, `default` or something similar.

dcbaker commented:
It would change the default on macOS, but not another platform.

Rather than a tri-state I'd prefer to have the default value on macos be true and everywhere else be false, which should be pretty simple using `compilers.get_base_compile_args`, does that seem reasonable?

xclaesse commented:
I think most linux distro change the default in gcc to enable PIE. At least on my Ubuntu everything in /usr/bin is PIE. So if anything, the default in meson should be to enable PIE. Do we even have a use-case where we want to explicitly disable PIE on platforms where it's enabled by default?

xclaesse commented:
```
xclaesse@xclaesse-x395:~$ cat test.c 
int main(){}
xclaesse@xclaesse-x395:~$ gcc test.c 
xclaesse@xclaesse-x395:~$ hardening-check a.out
a.out:
 Position Independent Executable: yes
(...)
```

xclaesse commented:
IMHO the pie option is meson should be understood as "force PIE". So when it's true you're sure to get a PIE exe, and when it's false you get the platform/compiler default which could be PIE.

bwidawsk commented:
> I think most linux distro change the default in gcc to enable PIE. At least on my Ubuntu everything in /usr/bin is PIE. So if anything, the default in meson should be to enable PIE. Do we even have a use-case where we want to explicitly disable PIE on platforms where it's enabled by default?

We (@marc-h38  and I) are using meson for a project that isn't yet public where we depend on this, and currently are forced to use a custom target to make it work.

> IMHO the pie option is meson should be understood as "force PIE". So when it's true you're sure to get a PIE exe, and when it's false you get the platform/compiler default which could be PIE.

It would be good if you renamed it then as pie: false doesn't work as advertised.

xclaesse commented:
Ok.... I guess that means we need a tristate as @jpakkane suggested.

xclaesse commented:
otoh, are there many platforms where PIE is not already enabled by default? Looks like all major linux distro enable it, MacOSX enable it, what about Windows?

I personally don't like much the tristate, it would be there only as backward compatibility reasons... dunno...

xclaesse commented:
Mmm, tested on a Fedora machine, looks like they don't enable PIE by default in gcc, but binaries in /usr/bin are PIE. Probably they enable it in packaging instead.

dcbaker commented:
I've updated to use a tristate, and I've added documentation changes.

I know that I'm changing the behavior of false, but At least me and @bwidawsk were genuinely surprised that false didn't mean `-no-pie`, and I suspect that we're not the only ones. The default value remains "compiler default", so I doubt anyone is going to be too surprised by this change.

dcbaker commented:
@jpakkane, @xclaesse

xclaesse commented on mesonbuild/build.py:
```diff
@@ -974,9 +974,15 @@ def _extract_pic_pie(self, kwargs, arg):
+        elif val == 'false':
+            val = False
+        if val not in {True, False, None}:
+            raise InvalidArguments('Argument {} to {!r} must one of "true", "false", "default"'.format(arg, self.name))
```
I don't think this part is correct.
- That's used to extract `pic` and `pie` from i.e. `executable()` keyword arguments. Possible values are bool or None, it must not be string 'true'.
- Do we want to support tristate for PIC too? So far this PR is only about PIE.
- You doc and changelog only mention change to the `b_pie` builtin option, not those keyword arguments. But I guess you're right that `executable(..., pie: false)` should produce `-no-pie` arg.

dcbaker replied:
I don't know about PIC, I've updated the documents to mention the change to the meson language

dcbaker replied:
I've also fixed the pie vs pic differences. Maybe it does make sense to do the same for PIC?

xclaesse replied:
We can handle the PIC case in another PR if anyone complains.

xclaesse commented on mesonbuild/build.py:
```diff
@@ -967,16 +967,32 @@ def process_kwargs(self, kwargs, environment):
+                val = None
+            elif val == 'true':
+                val = True
+            elif val == 'false':
```
still can't be a string `'true'` or `'false'` here. I think what you mean is:
```
val = kwargs.get(arg, None)
if val == 'default':
  val = None
elif not isinstance(val, bool):
  raise ...

```

dcbaker replied:
If the option comes from the `-Db_pie=` command line option it will be a string, because that's a combo.

xclaesse replied:
I don't think it can be string, otherwise current code wouldn't work. In Executable class it does: `kwargs['pie'] = environment.coredata.base_options['b_pie'].value`, and UserBooleanOption.value is a bool not string.

dcbaker replied:
It's not a UserBooleanOption in this series, it's a UserComboOption

xclaesse replied:
oohh, indeed, sorry completely missed the point here :D But that means that I would write `executable(..., pie : 'true')`, it's nitpicking but would be nice to forbid that syntax, no?

dcbaker replied:
Yeah, That probably should be done in the interpreter. I'll make some changes

dcbaker replied:
Okay, I've added an interpreter level check for strings other that `'default'` being passed to the `pie :` keyword argument. I've also added a FeatureNew for the default.

xclaesse replied:
Ok, I think that close this discussion here.

xclaesse commented on mesonbuild/interpreter.py:
```diff
@@ -4424,6 +4424,14 @@ def build_target_decorator_caller(self, node, args, kwargs):
+        if p is not None:
+            if p == 'default':
+                FeatureNew('Executable pie "default" argument', '0.54.0').use(self.subproject)
+            elif isinstance(p, str):
```
Rather `elif not isinstance(p, bool):`, no?

xclaesse commented on mesonbuild/compilers/compilers.py:
```diff
@@ -1038,12 +1040,17 @@ def get_pic_args(self):
+        it's false they'll disable it. IF it is None, then nothing will be
+        done, and you will get the compiler default.
+        """
+        return []
```
Shouldn't you raise exception here if enabled is not None? Otherwise compilers that does not override this method will silently, if I understand correctly.

dcbaker replied:
That's intentional, otherwise using a language without pie like D will always error if you set pie, which means you can't mix pie and non-pie languages and not use the default value. Which seems bad. I've added a test case in the latest version for that.

xclaesse commented on mesonbuild/compilers/mixins/islinker.py:
```diff
@@ -94,9 +94,9 @@ def get_allow_undefined_link_args(self) -> T.List[str]:
-        raise mesonlib.EnvironmentException(m.format(self.id))
+    def get_pie_link_args(self, enabled: T.Optional[bool]) -> T.List[str]:
+        """Get linker pie arguments."""
+        return []
```
Same here, removing the exception will make it silently ignore the pie value.

eli-schwartz commented on docs/markdown/Reference-manual.md:
```diff
@@ -646,7 +646,10 @@ be passed to [shared and static libraries](#library).
 - `d_debug` list of module debug identifiers set when compiling D sources
-- `pie` *(added 0.49.0)* build a position-independent executable
+- `pie` *(added 0.49.0)* build a position-independent executable. Before 0.54.0
+  `true` would force pie off, and `false` would use the compiler default. In
```
True would force pie *on*.
